### ISSUE TITLE
\Zend\Mail\Storage\Imap Invalid header value detected

### ISSUE THREAD

**Issue by Nirzol** [zendframework/zf2#7641]
Hello
Just test \Zend\Mail\Storage\Imap with the doc here : http://framework.zend.com/manual/current/en/modules/zend.mail.read.html

So simple case : 
```php
        $mail = new \Zend\Mail\Storage\Imap(array('host' => 'myhost',
            'user' => 'mylogin',
            'password' => 'mypassword'));
        
        $mail->getMessage(1);
```

I have this error : 
Invalid header value detected

But if I try another number with  : $mail->getMessage(7); 
it works.

use zf 2.5.2

**Comment by Maks3w** [comment]
Please provide the headers

**Comment by Nirzol** [comment]
Content-Type : multipart/mixed; boundary="----=_Part_18_1322065935.1387216214682"
Mime-Version : 1.0
X-Ms-Exchange-Organization-Authas : Anonymous
Return-Path : aip-demande@docapost-bpo.com
X-Ms-Exchange-Organization-Authsource : PENDRAGON.fougeres.campus
X-Ms-Exchange-Organization-Avstamp-Enterprise : 1.0
X-Ms-Exchange-Organization-Network-Message-Id : 849960a9-b2f5-4b7a-08db-08d0c89e6b3c
Message-Id : <2096357683.19.1387216214691.JavaMail.aipprd@waddle1>
Received : from GOUSTAN.fougeres.campus (172.22.6.50) by MORDRED.fougeres.campus (172.17.143.89) with Microsoft SMTP Server (TLS) id 15.0.712.24 via Mailbox Transport; Mon, 16 Dec 2013 18:50:55 +0100
Received : from PENDRAGON.fougeres.campus (172.17.143.88) by GOUSTAN.fougeres.campus (172.22.6.50) with Microsoft SMTP Server (TLS) id 15.0.712.24; Mon, 16 Dec 2013 18:50:18 +0100
Received : from mataram.parisdescartes.fr (172.17.143.79) by PENDRAGON.fougeres.campus (172.17.143.88) with Microsoft SMTP Server id 15.0.712.24 via Frontend Transport; Mon, 16 Dec 2013 18:50:19 +0100
Received : from localhost (unknown [192.168.253.9])	by mataram.parisdescartes.fr (Postfix) with ESMTP id 28F6E2FA345	for <me@me.fr>; Mon, 16 Dec 2013 18:50:19 +0100 (CET)
Received : from mataram.parisdescartes.fr ([192.168.253.4])	by localhost (fourmilier.univ-paris5.fr [192.168.253.9]) (amavisd-new, port 10024)	with ESMTP id g1qoJa7Ek2Np for <me@me.f>;	Mon, 16 Dec 2013 18:50:18 +0100 (CET)
Received : from mx1.extelia.fr (mx1.extelia.fr.jedeclare.info [194.5.108.155]) by mataram.parisdescartes.fr (Postfix) with ESMTP id 3B6DA2FA339	for <me@me.f>; Mon, 16 Dec 2013 18:50:18 +0100 (CET)
Received : from mx1.extelia.fr (localhost.localdomain [127.0.0.1])	by localhost (Postfix) with SMTP id B753B2484F5	for <me@me.f>; Mon, 16 Dec 2013 18:50:16 +0100 (CET)
Received : from smtp1.prod.extelia.fr (jarre.prod.extelia.fr [172.20.150.19]) by mx1.extelia.fr (Postfix) with ESMTP id C218D2486A6	for <me@me.f>; Mon, 16 Dec 2013 18:50:14 +0100 (CET)
Received : from smtp1.prod.extelia.fr (localhost.localdomain [127.0.0.1])	by localhost (Postfix) with SMTP id BCC3E2FA82	for <me@me.fr>; Mon, 16 Dec 2013 18:50:14 +0100 (CET)
Received : from waddle1 (smtp.jedeclare.info [172.20.27.250])	by smtp1.prod.extelia.fr (Postfix) with SMTP id A989D2FA7A	for <me@me.fr>; Mon, 16 Dec 2013 18:50:14 +0100 (CET)
X-Virus-Scanned : amavisd-new at univ-paris5.fr
Votre demande AIP - 12016193

**Comment by Maks3w** [comment]
Did you identify what header of them throw the exception?

**Comment by Nirzol** [comment]
Hum no sorry. How can I identify that ?

If you I can post the message source .

**Comment by Nirzol** [comment]
Maybe the trace can help ?

#0 .../vendor/zendframework/zend-mail/src/Header/GenericHeader.php(35): Zend\Mail\Header\GenericHeader::splitHeaderLine('Received: from ...')
#1 .../vendor/zendframework/zend-mail/src/Headers.php(231): Zend\Mail\Header\GenericHeader::fromString('Received: from ...')
#2 .../vendor/zendframework/zend-mail/src/Headers.php(95): Zend\Mail\Headers->addHeaderLine('Received: from ...')
#3 .../vendor/zendframework/zend-mime/src/Decode.php(141): Zend\Mail\Headers::fromString('MIME-Version: 1...', '\n')
#4 .../vendor/zendframework/zend-mail/src/Storage/Part.php(106): Zend\Mime\Decode::splitMessage('MIME-Version: 1...', 'MIME-Version: 1...', '')
#5 .../vendor/zendframework/zend-mail/src/Storage/Message.php(54): Zend\Mail\Storage\Part->__construct(Array)
#6 .../vendor/zendframework/zend-mail/src/Storage/Imap.php(118): Zend\Mail\Storage\Message->__construct(Array)
#7 .../module/Ent/src/Ent/Controller/IndexController.php(35): Zend\Mail\Storage\Imap->getMessage(1)
#8 .../vendor/zendframework/zend-mvc/src/Controller/AbstractActionController.php(82): Ent\Controller\IndexController->indexAction()
#9 [internal function]: Zend\Mvc\Controller\AbstractActionController->onDispatch(Object(Zend\Mvc\MvcEvent))
#10 .../vendor/zendframework/zend-eventmanager/src/EventManager.php(444): call_user_func(Array, Object(Zend\Mvc\MvcEvent))
#11 .../vendor/zendframework/zend-eventmanager/src/EventManager.php(205): Zend\EventManager\EventManager->triggerListeners('dispatch', Object(Zend\Mvc\MvcEvent), Object(Closure))
#12 .../vendor/zendframework/zend-mvc/src/Controller/AbstractController.php(118): Zend\EventManager\EventManager->trigger('dispatch', Object(Zend\Mvc\MvcEvent), Object(Closure))
#13 .../vendor/zendframework/zend-mvc/src/DispatchListener.php(93): Zend\Mvc\Controller\AbstractController->dispatch(Object(Zend\Http\PhpEnvironment\Request), Object(Zend\Http\PhpEnvironment\Response))
#14 [internal function]: Zend\Mvc\DispatchListener->onDispatch(Object(Zend\Mvc\MvcEvent))
#15 .../vendor/zendframework/zend-eventmanager/src/EventManager.php(444): call_user_func(Array, Object(Zend\Mvc\MvcEvent))
#16 .../vendor/zendframework/zend-eventmanager/src/EventManager.php(205): Zend\EventManager\EventManager->triggerListeners('dispatch', Object(Zend\Mvc\MvcEvent), Object(Closure))
#17 .../vendor/zendframework/zend-mvc/src/Application.php(314): Zend\EventManager\EventManager->trigger('dispatch', Object(Zend\Mvc\MvcEvent), Object(Closure))
#18 .../public/index.php(17): Zend\Mvc\Application->run()
#19 {main}

**Comment by Maks3w** [comment]
It helps. No we now its at least 1 of those Received headers. We don't know which of them yet.

**Comment by Nirzol** [comment]
Can I do something more ?

**Comment by Maks3w** [comment]
This look similar to some of the open issues about valid characters. If you can narrow to what character is throwing the exception will be agilize the triage of this issue.

**Comment by adamlundrigan** [comment]
#7524 also traces back to the same exception ("Invalid header value detected") triggered due to the restrictions imposed by `HeaderValue::isValid`(https://github.com/zendframework/zend-mail/blob/master/src/Header/HeaderValue.php#L73-L101), but with a different header (References).  My guess is your header contains a character that isn't acceptable in a structured header body ([see my comment here](https://github.com/zendframework/zf2/issues/7524#issuecomment-151694898))

**Comment by Maks3w** [comment]
This is related to https://github.com/zendframework/zend-mail/pull/28#issuecomment-150802504

**Comment by Nirzol** [comment]
I did some var dump in zf2 core and I think this line not pass in genericHeader.php The Line goes in that exception : 
```php
if (! HeaderValue::isValid(trim($parts[1]))) {
            throw new Exception\InvalidArgumentException('Invalid header value detected');
        }
```
And the part[1] is : 
```php
array (size=2)
  0 => string 'Received' (length=8)
  1 => string ' from localhost (unknown [192.168.253.9])	by mataram.parisdescartes.fr (Postfix) with ESMTP id 28F6E2FA345	for <eric.grondin@parisdescartes.fr>; Mon, 16 Dec 2013 18:50:19 +0100 (CET)' (length=182)
```

**Comment by Nirzol** [comment]
The bad character is the 41. The blank one just after : 
 from localhost (unknown [192.168.253.9])

the ord function return 9

**Comment by Maks3w** [comment]
Thx. Then will fixed with #28.

I'll close it because duplicates that issue.
